include/ddc/kernels/splines/matrix_sparse.hpp

```diff
@@ -0,0 +1,255 @@
+        Kokkos::realloc(Kokkos::WithoutInitializing, data, data_mat_.nonzeros.size());
+        Kokkos::deep_copy(
+                rows,
+                Kokkos::View<
+                        int*,
+                        Kokkos::HostSpace,
+                        Kokkos::MemoryTraits<Kokkos::Unmanaged>>(data_mat->get_row_ptrs(), m + 1));
+        Kokkos::deep_copy(
+                cols,
+                Kokkos::View<int*, Kokkos::HostSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
+                        data_mat->get_col_idxs(),
+                        data_mat->get_num_stored_elements()));
+        Kokkos::deep_copy(
+                data,
+                Kokkos::View<double*, Kokkos::HostSpace, Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
+                        data_mat->get_values(),
+                        data_mat->get_num_stored_elements()));
+        return 0;
+    }
+    virtual int solve_inplace_method(double* b, char transpose, int n_equations) const override
+    {
+        std::shared_ptr<gko::Executor> gko_exec = create_gko_exec<ExecSpace>();
+        auto data_mat = gko::share(
+                to_gko_mat(data.data(), rows.size() - 1, cols.size(), gko_exec->get_master()));
+        auto data_mat_gpu = gko::share(gko::clone(gko_exec, data_mat));
+        Kokkos::View<
+                double**,
+                Kokkos::LayoutRight,
+                ExecSpace,
+                Kokkos::MemoryTraits<Kokkos::Unmanaged>>
+                b_view(b, n, n_equations);
+
+        const int n_seq_chunks = n_equations / cols_per_par_chunk / par_chunks_per_seq_chunk + 1;
+        const int par_chunks_per_last_seq_chunk
+                = (n_equations % (cols_per_par_chunk * par_chunks_per_seq_chunk))
+                          / cols_per_par_chunk
+                  + 1;
+        const int cols_per_last_par_chunk
+                = (n_equations % (cols_per_par_chunk * par_chunks_per_seq_chunk * n_seq_chunks))
+                  % cols_per_par_chunk;
+
+        Kokkos::View<double***, Kokkos::LayoutRight, ExecSpace>
+                b_buffer("b_buffer", par_chunks_per_seq_chunk, n, cols_per_par_chunk);
+        // Last par_chunk of last seq_chunk do not have same number of columns than the others. To get proper layout (because we passe the pointers to Ginkgo), we need a dedicated allocation
+        Kokkos::View<double**, Kokkos::LayoutRight, ExecSpace>
+                b_last_buffer("b_last_buffer", n, cols_per_last_par_chunk);
+
+        // Sequential loop
+        for (int i = 0; i < n_seq_chunks; i++) {
+            int n_par_chunks_in_seq_chunk = i < n_seq_chunks - 1 ? par_chunks_per_seq_chunk
+                                                                 : par_chunks_per_last_seq_chunk;
+            // Parallel loop
+            Kokkos::parallel_for(
+                    Kokkos::RangePolicy<
+                            Kokkos::DefaultHostExecutionSpace>(0, n_par_chunks_in_seq_chunk),
```

Comment:
I have a few questions regarding this loop:
- what about a pure OpenMP run ? In the current implementation I think each OpenMP thread will open a new parallel region inside Ginkgo ?
- Have you checked that it is valid to call these Kokkos functions inside a `parallel_for`, be it CPU ?
- In terms of performance, can you show using NVIDIA tools that we have a benefit of using OpenMP + CUDA ?

Reply:
I realized Ginkgo works in Serial with a multiple-rhs call in OpenMP. Not sure why however. So calling Ginkgo in parallel region seems necessary in OpenMP configuration to benefit from parallelism. GPU is different.

I just checked the tests results are good and each thread acts on a different dataset so no race condition happens. Also we can assume those basic Kokkos functions to be thread-safe, no ?

I dont think we have benefit of using OpenMP+CUDA, reason why par_chunks_per_seq_chunk=1 in CUDA configuration.